File: src/main/kotlin/magicbook/gtlitecore/api/metatileentity/multiblock/ModuleMultiblockBase.java
package magicbook.gtlitecore.api.metatileentity.multiblock;

import com.cleanroommc.modularui.factory.PosGuiData;
import com.cleanroommc.modularui.screen.ModularPanel;
import com.cleanroommc.modularui.value.sync.PanelSyncManager;
import gregtech.api.capability.GregtechDataCodes;
import gregtech.api.capability.GregtechTileCapabilities;
import gregtech.api.capability.IControllable;
import gregtech.api.capability.IEnergyContainer;
import gregtech.api.capability.IWorkable;
import gregtech.api.capability.impl.EnergyContainerHandler;
import gregtech.api.metatileentity.multiblock.IMultiblockPart;
import gregtech.api.metatileentity.multiblock.MultiblockWithDisplayBase;
import gregtech.api.pattern.BlockPattern;
import gregtech.api.pattern.PatternMatchContext;
import gregtech.api.util.GTUtility;
import gregtech.client.renderer.ICubeRenderer;
import lombok.Getter;
import lombok.Setter;
import magicbook.gtlitecore.api.capability.IModuleProvider;
import magicbook.gtlitecore.api.capability.IModuleReceiver;
import magicbook.gtlitecore.api.gui.GTLiteMuiTextures;
import magicbook.gtlitecore.api.gui.factory.MultiblockUIBuilder;
import magicbook.gtlitecore.api.gui.factory.MultiblockUIFactory;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import org.jetbrains.annotations.NotNull;

import static magicbook.gtlitecore.api.utils.GTLiteValues.SECOND;

public abstract class ModuleMultiblockBase extends MultiblockWithDisplayBase implements IModuleReceiver, IWorkable, IControllable
{

    private final int tier;
    private final int moduleTier;
    private final int minCasingTier;

    @Getter
    @Setter
    protected IModuleProvider moduleProvider;

    protected IEnergyContainer energyContainer;
    protected final long energyConsumed;

    protected boolean isActive;
    @Setter
    @Getter
    protected int maxProgress;
    protected int progressTime = 0;
    protected boolean isWorkingEnabled = false;

    /**
     * @param tier          The voltage tier of this mte.
     * @param moduleTier    The inner tier of the module.
     * @param minCasingTier The minimum casing tier of this module required, this is useful for some
     *                      tiered status predicate.
     */
    public ModuleMultiblockBase(ResourceLocation metaTileEntityId, int tier,
                                int moduleTier, int minCasingTier)
    {
        super(metaTileEntityId);
        this.tier = tier;
        this.moduleTier = moduleTier;
        this.minCasingTier = minCasingTier;
        this.energyConsumed = (long) (Math.pow(4, this.tier + 2) / 2);
        this.energyContainer = new EnergyContainerHandler(this,
                (long) (160_008_000 * Math.pow(4, this.tier - 9)), this.energyConsumed,
                1, 0, 0);
    }

    @Override
    protected void formStructure(PatternMatchContext context)
    {
        super.formStructure(context);
        initializeAbilities();
    }

    protected abstract void initializeAbilities();

    @Override
    public void checkStructurePattern()
    {
        super.checkStructurePattern();
        if (getModuleProvider() != null)
        {
            if (getModuleProvider().getCasingTier() >= this.minCasingTier)
                super.checkStructurePattern();
        }
    }

    @Override
    public void invalidateStructure()
    {
        super.invalidateStructure();
        setModuleProvider(null);
    }

    @NotNull
    @Override
    protected abstract BlockPattern createStructurePattern();

    @Override
    protected void updateFormedValid()
    {
        if (getOffsetTimer() % SECOND == 0 && getModuleProvider() != null)
        {
            if (energyContainer.getEnergyCapacity() != energyContainer.getEnergyStored()
                    && getModuleProvider().getSubEnergyContainer().getEnergyStored() > energyConsumed * SECOND)
            {
                long simulate = energyContainer.getEnergyCapacity() - energyContainer.getEnergyStored();
                energyContainer.addEnergy(simulate);
            }
        }
        else if (getModuleProvider() == null)
        {
            setWorkingEnabled(false);
        }
    }

    public IEnergyContainer getEnergyContainer()
    {
        if (getModuleProvider() == null || getModuleProvider().getSubEnergyContainer() == null)
        {
            return new EnergyContainerHandler(this, 0, 0, 0, 0, 0);
        }
        else
        {
            return energyContainer;
        }
    }

    @Override
    public <T> T getCapability(Capability<T> capability, EnumFacing side)
    {
        if (capability == GregtechTileCapabilities.CAPABILITY_WORKABLE)
            return GregtechTileCapabilities.CAPABILITY_WORKABLE.cast(this);
        if (capability == GregtechTileCapabilities.CAPABILITY_CONTROLLABLE)
            return GregtechTileCapabilities.CAPABILITY_CONTROLLABLE.cast(this);
        return super.getCapability(capability, side);
    }

    @Override
    public NBTTagCompound writeToNBT(NBTTagCompound data)
    {
        super.writeToNBT(data);
        data.setInteger("progressTime", progressTime);
        data.setInteger("maxProgress", maxProgress);
        data.setBoolean("isActive", isActive);
        data.setBoolean("isWorkingEnabled", isWorkingEnabled);
        return data;
    }

    @Override
    public void readFromNBT(NBTTagCompound data)
    {
        super.readFromNBT(data);
        progressTime = data.getInteger("progressTime");
        maxProgress = data.getInteger("maxProgress");
        isActive = data.getBoolean("isActive");
        isWorkingEnabled = data.getBoolean("isWorkingEnabled");
    }

    @Override
    public void writeInitialSyncData(PacketBuffer buf)
    {
        super.writeInitialSyncData(buf);
        buf.writeInt(progressTime);
        buf.writeInt(maxProgress);
        buf.writeBoolean(isActive);
        buf.writeBoolean(isWorkingEnabled);
    }

    @Override
    public void receiveInitialSyncData(PacketBuffer buf)
    {
        super.receiveInitialSyncData(buf);
        progressTime = buf.readInt();
        maxProgress = buf.readInt();
        setActive(buf.readBoolean());
        setWorkingEnabled(buf.readBoolean());
    }

    @Override
    public void receiveCustomData(int dataId, PacketBuffer buf)
    {
        super.receiveCustomData(dataId, buf);
        if (dataId == GregtechDataCodes.WORKABLE_ACTIVE)
        {
            setActive(buf.readBoolean());
            scheduleRenderUpdate();
        }
        else if (dataId == GregtechDataCodes.WORKING_ENABLED)
        {
            setWorkingEnabled(buf.readBoolean());
            scheduleRenderUpdate();
        }
    }

    @Override
    public boolean isActive()
    {
        return isActive && isWorkingEnabled();
    }

    public void setActive(boolean active)
    {
        if (isActive != active)
        {
            isActive = active;
            markDirty();
            if (getWorld() != null && !getWorld().isRemote)
                writeCustomData(GregtechDataCodes.WORKABLE_ACTIVE, buf -> buf.writeBoolean(active));
        }
    }

    @Override
    public int getProgress()
    {
        return progressTime;
    }

    @Override
    public boolean isWorkingEnabled()
    {
        return isWorkingEnabled;
    }

    @Override
    public void setWorkingEnabled(boolean workingEnabled)
    {
        initializeAbilities();
        isWorkingEnabled = workingEnabled;
        markDirty();
        if (getWorld() != null && !getWorld().isRemote)
            writeCustomData(GregtechDataCodes.WORKING_ENABLED, buf -> buf.writeBoolean(workingEnabled));
    }

    protected boolean drainEnergy(boolean simulate, long energy)
    {
        long result = energyContainer.getEnergyStored() - energy;
        if (result >= 0L && result <= energyContainer.getEnergyCapacity())
        {
            if (!simulate)
                energyContainer.changeEnergy(-energy);
            return true;
        }
        return false;
    }

    protected boolean drainEnergy(boolean simulate)
    {
        long result = energyContainer.getEnergyStored() - energyContainer.getInputVoltage();
        if (result >= 0L && result <= energyContainer.getEnergyCapacity())
        {
            if (!simulate)
                energyContainer.changeEnergy(-energyContainer.getInputVoltage());
            return true;
        }
        return false;
    }

    @Override
    public String getDisplayCountName()
    {
        return getMetaName() + ".display_count";
    }

    public int getProgressPercent()
    {
        return (int) ((1.0F * getProgress() / getMaxProgress()) * 100);
    }

    @SuppressWarnings("UnstableApiUsage")
    @Override
    public boolean usesMui2()
    {
        return true;
    }

    @Override
    public ModularPanel buildUI(PosGuiData guiData, PanelSyncManager guiSyncManager)
    {
        return new MultiblockUIFactory(this)
                .configureDisplayText(this::configureDisplayText)
                .configureWarningText(this::configureWarningText)
                .configureErrorText(this::configureErrorText)
                .buildUI(guiData, guiSyncManager, GTLiteMuiTextures.DISPLAY);
    }

    protected abstract void configureDisplayText(MultiblockUIBuilder builder);

    protected abstract void configureWarningText(MultiblockUIBuilder builder);

    protected abstract void configureErrorText(MultiblockUIBuilder builder);

    @SideOnly(Side.CLIENT)
    @NotNull
    @Override
    protected abstract ICubeRenderer getFrontOverlay();

    @SideOnly(Side.CLIENT)
    @Override
    public abstract ICubeRenderer getBaseTexture(IMultiblockPart sourcePart);

    @Override
    public boolean hasMaintenanceMechanics()
    {
        return false;
    }

    @Override
    public void sentWorkingDisabled()
    {
        setWorkingEnabled(false);
    }

    @Override
    public void sentWorkingEnabled()
    {
        setWorkingEnabled(true);
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/api/metatileentity/multiblock/ModuleMultiblockBase.kt
package gregtechlite.gtlitecore.api.metatileentity.multiblock

import gregtech.api.capability.GregtechDataCodes.WORKABLE_ACTIVE
import gregtech.api.capability.GregtechDataCodes.WORKING_ENABLED
import gregtech.api.capability.GregtechTileCapabilities.CAPABILITY_CONTROLLABLE
import gregtech.api.capability.GregtechTileCapabilities.CAPABILITY_WORKABLE
import gregtech.api.capability.IControllable
import gregtech.api.capability.IEnergyContainer
import gregtech.api.capability.IWorkable
import gregtech.api.capability.impl.EnergyContainerHandler
import gregtech.api.capability.impl.EnergyContainerList
import gregtech.api.metatileentity.multiblock.IMultiblockPart
import gregtech.api.metatileentity.multiblock.MultiblockWithDisplayBase
import gregtech.api.pattern.BlockPattern
import gregtech.api.pattern.PatternMatchContext
import gregtech.client.renderer.ICubeRenderer
import gregtechlite.gtlitecore.api.SECOND
import gregtechlite.gtlitecore.api.capability.ModuleProvider
import gregtechlite.gtlitecore.api.capability.ModuleReceiver
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.PacketBuffer
import net.minecraft.util.EnumFacing
import net.minecraft.util.ResourceLocation
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import kotlin.math.min
import kotlin.math.pow

abstract class ModuleMultiblockBase(metaTileEntityId: ResourceLocation,
                                    protected val tier: Int,
                                    protected val moduleTier: Int,
                                    protected val minCasingTier: Int)
    : MultiblockWithDisplayBase(metaTileEntityId), ModuleReceiver, IWorkable, IControllable
{
    override var moduleProvider: ModuleProvider? = null

    override val displayCountName: String
        get() = "$metaName.display_count"

    @JvmField
    protected var energyContainer: IEnergyContainer
    protected val energyConsumed = (4.0.pow((this.tier + 2).toDouble()) / 2).toLong()

    @JvmField
    protected var isActive: Boolean = false

    @JvmField
    protected var maxProgress: Int = 0

    @JvmField
    protected var progress: Int = 0

    val progressPercent: Int
        get() = ((1.0f * progress / maxProgress) * 100).toInt()

    @JvmField
    protected var isWorkingEnabled: Boolean = false

    /**
     * @param tier          The voltage tier of this mte.
     * @param moduleTier    The inner tier of the module.
     * @param minCasingTier The minimum casing tier of this module required, this is useful for some
     *                      tiered status predicate.
     */
    init
    {
        energyContainer = EnergyContainerHandler(this, (160008000L * 4.0.pow((tier - 9).toDouble())).toLong(),
            energyConsumed, 1, 0, 0)
    }

    override fun formStructure(context: PatternMatchContext?)
    {
        super.formStructure(context)
        initializeAbilities()
    }

    protected abstract fun initializeAbilities()

    override fun checkStructurePattern()
    {
        super.checkStructurePattern()
        moduleProvider?.casingTier?.also {
            if (it >= minCasingTier)
            {
                super.checkStructurePattern()
            }
        }
    }

    override fun invalidateStructure()
    {
        super.invalidateStructure()
        moduleProvider = null
    }

    abstract override fun createStructurePattern(): BlockPattern

    override fun updateFormedValid()
    {
        if (offsetTimer % SECOND == 0L && moduleProvider != null)
        {
            if (energyContainer.energyCapacity != energyContainer.energyStored
                && moduleProvider!!.subEnergyContainer!!.energyStored > energyConsumed * SECOND)
            {
                val maxModuleReceive = energyContainer.energyCapacity - energyContainer.energyStored
                val energyDrained = min(moduleProvider!!.subEnergyContainer!!.energyStored, maxModuleReceive)

                moduleProvider!!.subEnergyContainer!!.removeEnergy(energyDrained)
                energyContainer.addEnergy(energyDrained)
            }
        }
        else if (moduleProvider == null)
        {
            isWorkingEnabled = false
        }
    }

    fun getEnergyContainer(): IEnergyContainer?
        = if (moduleProvider?.subEnergyContainer == null) EnergyContainerHandler(this, 0, 0, 0, 0, 0) else energyContainer

    fun getCombinedEnergyContainer(): IEnergyContainer
    = if (moduleProvider?.subEnergyContainer == null) EnergyContainerHandler(this, 0, 0, 0, 0, 0)
        else EnergyContainerList(listOf(moduleProvider!!.subEnergyContainer, energyContainer))

    override fun <T : Any> getCapability(capability: Capability<T>, side: EnumFacing?): T?
    {
        if (capability === CAPABILITY_WORKABLE) return CAPABILITY_WORKABLE.cast(this)
        if (capability === CAPABILITY_CONTROLLABLE) return CAPABILITY_CONTROLLABLE.cast(this)
        return super.getCapability(capability, side)
    }

    override fun writeToNBT(data: NBTTagCompound): NBTTagCompound
    {
        super.writeToNBT(data)
        data.setInteger("progressTime", progress)
        data.setInteger("maxProgress", maxProgress)
        data.setBoolean("isActive", isActive)
        data.setBoolean("isWorkingEnabled", isWorkingEnabled)
        return data
    }

    override fun readFromNBT(data: NBTTagCompound)
    {
        super.readFromNBT(data)
        progress = data.getInteger("progressTime")
        maxProgress = data.getInteger("maxProgress")
        isActive = data.getBoolean("isActive")
        isWorkingEnabled = data.getBoolean("isWorkingEnabled")
    }

    override fun writeInitialSyncData(buf: PacketBuffer)
    {
        super.writeInitialSyncData(buf)
        buf.writeInt(progress)
        buf.writeInt(maxProgress)
        buf.writeBoolean(isActive)
        buf.writeBoolean(isWorkingEnabled)
    }

    override fun receiveInitialSyncData(buf: PacketBuffer)
    {
        super.receiveInitialSyncData(buf)
        progress = buf.readInt()
        maxProgress = buf.readInt()
        isActive = buf.readBoolean()
        isWorkingEnabled = buf.readBoolean()
    }

    override fun receiveCustomData(dataId: Int, buf: PacketBuffer)
    {
        super.receiveCustomData(dataId, buf)
        when (dataId)
        {
            WORKABLE_ACTIVE ->
            {
                setActive(buf.readBoolean())
                scheduleRenderUpdate()
            }
            WORKING_ENABLED ->
            {
                isWorkingEnabled = buf.readBoolean()
                scheduleRenderUpdate()
            }
        }
    }

    override fun isActive(): Boolean = isActive && isWorkingEnabled

    fun setActive(active: Boolean)
    {
        if (isActive != active)
        {
            isActive = active
            markDirty()
            if (world != null && !world.isRemote)
            {
                writeCustomData(WORKABLE_ACTIVE) { it.writeBoolean(active) }
            }
        }
    }

    override fun getProgress(): Int = progress

    protected fun drainEnergy(simulate: Boolean, energy: Long = energyContainer.inputVoltage): Boolean
    {
        val container = getCombinedEnergyContainer()
        val result = container.energyStored - energy
        if (result >= 0L && result <= container.energyCapacity)
        {
            if (!simulate) container.changeEnergy(-energy)
            return true
        }
        return false
    }

    @SideOnly(Side.CLIENT)
    abstract override fun getFrontOverlay(): ICubeRenderer

    @SideOnly(Side.CLIENT)
    abstract override fun getBaseTexture(sourcePart: IMultiblockPart?): ICubeRenderer?

    override fun hasMaintenanceMechanics() = false

    override fun sentWorkingDisabled()
    {
        isWorkingEnabled = false
    }

    override fun sentWorkingEnabled()
    {
        isWorkingEnabled = true
    }

    override fun isWorkingEnabled(): Boolean = this.isWorkingEnabled

    override fun setWorkingEnabled(workingEnabled: Boolean)
    {
        initializeAbilities()
        isWorkingEnabled = workingEnabled
        markDirty()
        if (world != null && !world.isRemote)
        {
            writeCustomData(WORKING_ENABLED) { it.writeBoolean(isWorkingEnabled) }
        }
    }

    override fun getMaxProgress() = maxProgress
}
